File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
def solution(maps):
    n = len(maps)
    m = len(maps[0])
    move = [[-1,0], [1,0], [0,-1],[0,1]]
    
    def bfs(x, y):
        que = deque()
        que.append((x, y))
        while que:
            x, y = que.popleft()
            for d_x, d_y in move:
                next_x = x + d_x
                next_y = y + d_y
                if 0 <= next_x < n and 0 <= next_y < m and maps[next_x][next_y] == 1:
                    maps[next_x][next_y] += maps[x][y]
                    que.append((next_x, next_y))
        return -1 if maps[n-1][m-1] == 1 else maps[n-1][m-1]
    answer = bfs(0, 0)
    
    return answer

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    const move = [[1, 0], [-1, 0], [0, 1], [0, -1]]
    const X = maps.length
    const Y = maps[0].length
    const BFS = (x, y) => {
        const stack = [[x, y]]
        
        while (stack.length > 0) {
            const current = stack.shift()
            for (let i = 0; i < move.length; i++) {
                const nextX = current[0] + move[i][0]
                const nextY = current[1] + move[i][1]
                
                if (nextX >= 0 && nextY >= 0 && nextX < X && nextY < Y && maps[nextX][nextY] === 1) {
                    maps[nextX][nextY] += maps[current[0]][current[1]]
                    stack.push([nextX, nextY])
                }
            }
        }
        return maps[X - 1][Y - 1] === 1 ? -1 : maps[X - 1][Y - 1]
    }
    
    
    return BFS(0, 0)
}